tooling/Microsoft.VisualStudio.BlazorExtension/AutoRebuild/AutoRebuildService.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.VisualStudio.Shell.Interop;
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipes;
+using System.Threading;
+using System.Threading.Tasks;
+using Package = Microsoft.VisualStudio.Shell.Package;
+using ThreadHelper = Microsoft.VisualStudio.Shell.ThreadHelper;
+
+namespace Microsoft.VisualStudio.BlazorExtension
+{
+    /// <summary>
+    /// The counterpart to VSForWindowsRebuildService.cs in the Blazor.Server project.
+    /// Listens for named pipe connections and rebuilds projects on request.
+    /// </summary>
+    internal class AutoRebuildService
+    {
+        private const int _protocolVersion = 1;
+        private readonly BuildEventsWatcher _buildEventsWatcher;
+        private readonly string _pipeName;
+
+        public AutoRebuildService(BuildEventsWatcher buildEventsWatcher)
+        {
+            _buildEventsWatcher = buildEventsWatcher ?? throw new ArgumentNullException(nameof(buildEventsWatcher));
+            _pipeName = $"BlazorAutoRebuild\\{Process.GetCurrentProcess().Id}";
+        }
+
+        public void Listen()
+        {
+            AddBuildServiceNamedPipeServer();
+        }
+
+        private void AddBuildServiceNamedPipeServer()
+        {
+            Task.Factory.StartNew(async () =>
+            {
+                try
+                {
+                    using (var serverPipe = new NamedPipeServerStream(_pipeName, PipeDirection.InOut, NamedPipeServerStream.MaxAllowedServerInstances))
```

Comment:
There's a problem here with permissions, and I'm not sure how hardcore the VS folk are about this.

This is an escalation of privilege vector. If VS is running as admin, a non-admin can connect to this an cause VS to rebuild arbitrary projects . Or another user. 

The fix for it is to do this: https://github.com/dotnet/roslyn/blob/14aed138a01c448143b9acf0fe77a662e3dfe2f4/src/Compilers/Server/VBCSCompiler/NamedPipeClientConnection.cs#L93

and this: https://github.com/dotnet/roslyn/blob/14aed138a01c448143b9acf0fe77a662e3dfe2f4/src/Compilers/Server/VBCSCompiler/NamedPipeClientConnection.cs#L160

Reply:
Thanks for pointing this out!

If I'm understanding this correctly, it seems like we only need to do the first one of those things now. The second one was a workaround for .NET Core limitations:

> // The overload of NamedPipeServerStream with the PipeAccessRule
> // parameter was removed in netstandard

... but that limitation seems to have done. The constructor is there, probably since .NET Core 2.0.

Do you think the following is therefore sufficient? https://github.com/aspnet/Blazor/pull/531/files#diff-e51aea7430ab7fad3d28c4afdb59db00R53

Reply:
It would be good to check if the user is at the same permission level. I don't think ACLs check this. Some users run VS as admin, and in theory any process running as a normal user could send data to our pipe.

Reply:
Note that they also compare SIDs which covers the per-user thing. I'm just adding the check for elevation.